### ISSUE TITLE
[Bug][A5] Pending mixed C2V producer corrupts a running Vector task’s UB

### ISSUE THREAD

**Issue by high-cloud** [hw-native-sys/pypto#2829]
## Background

DeepSeek V4.1 C1A prefill in pypto-lib intermittently fails TP4 Top-K validation because the **scores are corrupted before sorting**. We reduced the failure to a **single-device, three-task PyPTO program** with no model code, external weights, generated artifacts, or pypto-lib dependency. Its only dependencies are PyPTO and PyTorch.

On the pinned A5 environment below, the standalone program failed **5/5 runs**. Adding one Vector-completion dependency, without changing any arithmetic, passed **5/5 runs**. All test values have exact expected results; comparison uses zero tolerance.

### Reproduction environment

| Component | Revision / version |
| --- | --- |
| PyPTO | `1deae896dfe2c77e6302df1eef712887877264da` (detached, V4.1 CI pin) |
| simpler | `39ce891dbb3f665e72e99b4a1387b47012fb18bd` (detached) |
| PTO ISA | `5a4f74cbf627d4aac2e0ce10d5e0d8b118343265` (detached) |
| PTOAS | `0.61` |
| CANN | `9.2.0` |
| Python / PyTorch | `3.11.16` / `2.10.0+cpu` |
| Device / host | A5 hardware, one device / Linux aarch64 |
| Original model checkout | pypto-lib `cd2b7f808b61b272144b453ed6de212d5f9e7fdb` |

The simpler gitlink, PTOAS version, and simpler-managed PTO ISA checkout all match the PyPTO pin chain. No runtime or ISA patches were applied. This report does not establish behavior on newer revisions or on a simulator.

## Component

Backend / runtime integration: **A5 MIX-PENDING scheduling and the local no-split C2V startup protocol**. Filing in PyPTO to track the generated mixed-kernel contract across simpler and PTO ISA; the ownership boundary is described below.

## Steps to reproduce

Save the complete script below as `repro_a5_c2v_ub_overlap.py`. Activate the environment containing the revisions above and the CANN environment; configure `PTOAS_ROOT` as usual. No pypto-lib checkout or input files are required.

```bash
python repro_a5_c2v_ub_overlap.py --device 0 --repeat 5
python repro_a5_c2v_ub_overlap.py --device 0 --repeat 5 --ordered
```

On a machine using task-submit, run each command through its device allocator, for example:

```bash
task-submit --device auto --device-num 1 --max-time 0 --timeout 0 --run \
  'python repro_a5_c2v_ub_overlap.py --device "$TASK_DEVICE" --repeat 5'
```

<details>
<summary>Complete standalone reproducer: repro_a5_c2v_ub_overlap.py</summary>

```python
"""Reproduce A5 local C2V writes into an unrelated running Vector task's UB.

Requires only PyPTO and PyTorch. Run on one A5 device with --repeat 5.
Add --ordered to prevent the mixed projection from overlapping Vector work.
All input values and expected outputs are exact; no tolerance is involved.
"""
import argparse
import pypto.language as pl
import torch
from pypto.runtime import RunConfig

M = 16
N = 32
K = 7168
K_TILE = 256
V = 4096
WORKERS = 64
ITERATIONS = 8192


def make_case(ordered):
    @pl.jit
    def repro(
        x: pl.Tensor[[M, K], pl.BF16],
        w: pl.Tensor[[K, N], pl.BF16],
        v: pl.Tensor[[WORKERS, V], pl.FP32],
        y: pl.Out[pl.Tensor[[M, N], pl.BF16]],
        z: pl.Out[pl.Tensor[[WORKERS, V], pl.FP32]],
        gate: pl.Out[pl.Tensor[[M, N], pl.FP32]],
    ):
        with pl.spmd(WORKERS, name_hint="independent_vector") as vector_tid:
            block = pl.tile.get_block_idx()
            value = pl.load(v, [block, 0], [1, V])
            for step in pl.range(ITERATIONS):
                value = pl.add(value, 1.0)
            pl.store(value, [block, 0], z)
        # Delay readiness of the mixed task while independent Vector work runs.
        with pl.spmd(1, name_hint="cube_gate") as gate_tid:
            block = pl.tile.get_block_idx()
            acc = pl.create_tensor([M, N], dtype=pl.FP32)
            for kb in pl.range(K // K_TILE):
                a = x[:, kb * K_TILE:(kb + 1) * K_TILE]
                b = w[kb * K_TILE:(kb + 1) * K_TILE, :]
                acc = pl.matmul_acc(acc, a, b, init_cond=(kb == 0))
            gate[block:block + M, :] = acc
        deps = pl.array.create(1, pl.TASK_ID)
        deps[0] = gate_tid
        if ordered:
            deps[0] = vector_tid
        with pl.spmd(1, name_hint="mixed_projection", deps=[gate_tid, deps[0]]):
            block = pl.tile.get_block_idx()
            accumulator = pl.create_tensor([M, N], dtype=pl.FP32)
            for kb in pl.range(K // K_TILE):
                a = x[:, kb * K_TILE:(kb + 1) * K_TILE]
                b = w[kb * K_TILE:(kb + 1) * K_TILE, :]
                accumulator = pl.matmul_acc(accumulator, a, b, init_cond=(kb == 0))
            projected = pl.cast(accumulator, pl.BF16, mode="rint")
            scaled = pl.mul(pl.cast(projected, pl.FP32), 1.0 / 64)
            y[block:block + M, :] = pl.cast(scaled, pl.BF16, mode="rint")
        return y, z, gate
    return repro


def check(name, actual, expected):
    bad = actual != expected
    count = int(bad.sum())
    if count:
        i = bad.nonzero()[0].tolist()
        print(f"{name}: {count}/{actual.numel()} unequal; first={i}, "
              f"actual={actual[tuple(i)].item()}, expected={expected[tuple(i)].item()}", flush=True)
    return count == 0


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--device", type=int, default=0)
    parser.add_argument("--ordered", action="store_true")
    parser.add_argument("--repeat", type=int, default=5)
    args = parser.parse_args()
    if args.repeat < 1:
        parser.error("--repeat must be positive")
    x = torch.ones(M, K, dtype=torch.bfloat16)
    w = torch.ones(K, N, dtype=torch.bfloat16)
    v = torch.ones(WORKERS, V, dtype=torch.float32)
    y = torch.empty(M, N, dtype=torch.bfloat16)
    z = torch.empty(WORKERS, V, dtype=torch.float32)
    gate = torch.empty(M, N, dtype=torch.float32)
    expected_y = torch.full_like(y, K / 64)
    expected_z = v + ITERATIONS
    kernel = make_case(args.ordered)
    config = RunConfig(platform="a5", device_id=args.device)
    failures = 0
    for iteration in range(args.repeat):
        kernel(x, w, v, y, z, gate, config=config)
        y_ok = check("projection", y, expected_y)
        z_ok = check("independent_vector", z, expected_z)
        gate_ok = check("gate", gate, torch.full_like(gate, K))
        passed = y_ok and z_ok and gate_ok
        failures += not passed
        print(f"iteration={iteration} ordered={args.ordered} {'PASS' if passed else 'FAIL'}", flush=True)
    print(f"{failures}/{args.repeat} failed", flush=True)
    raise SystemExit(1 if failures else 0)
```

</details>

## Expected behavior

For both modes, every projection element is exactly `7168 / 64 = 112`; every independent Vector element is exactly `1 + 8192 = 8193`; every pure-Cube gate element is exactly `7168`.

The Vector task and mixed projection access disjoint GM input/output tensors. Their arithmetic does not require a dependency. A pure-Cube gate delays the mixed task's readiness while independent Vector work is running. `--ordered` adds the Vector completion to the mixed task's dependencies and preserves its Cube-gate dependency.

## Actual behavior

Unordered, final script, five invocations in one process:

```text
projection: 512/512 unequal; first=[0, 0], actual=232.0, expected=112.0
independent_vector: 512/262144 unequal; first=[48, 0], actual=14819.0, expected=8193.0
iteration=0 ordered=False FAIL
...
5/5 failed
```

The precise corrupted values vary with scheduling. All five invocations corrupt all 512 projection elements; the independent Vector output is also corrupted. The pure-Cube gate output passes.

Ordered control, same script and arithmetic:

```text
iteration=0 ordered=True PASS
iteration=1 ordered=True PASS
iteration=2 ordered=True PASS
iteration=3 ordered=True PASS
iteration=4 ordered=True PASS
0/5 failed
```

## Diagnosis and evidence

The generated mixed projection uses `TPipe<0, Direction::DIR_C2V, 2048, 2, 2, true>` with `TileSplitAxis::TILE_NO_SPLIT`; its local consumer buffer starts at UB offset 0. The independent Vector task also has live data at UB offset 0. Sharing that address across **sequential tasks** is legitimate; it becomes unsafe when the pending mixed task's Cube producer writes there before its Vector consumer owns the core.

Relevant pinned source:

- [simpler scheduler_types.h, classify_mix_cluster](https://github.com/hw-native-sys/simpler/blob/39ce891dbb3f665e72e99b4a1387b47012fb18bd/src/a5/runtime/tensormap_and_ringbuffer/runtime/scheduler/scheduler_types.h#L347): a MIX cluster may be PENDING when some active cores are busy but pending slots are free.
- [simpler scheduler_dispatch.cpp](https://github.com/hw-native-sys/simpler/blob/39ce891dbb3f665e72e99b4a1387b47012fb18bd/src/a5/runtime/tensormap_and_ringbuffer/runtime/scheduler/scheduler_dispatch.cpp#L226): idle AIC starts in its running slot while the busy paired AIV receives the mixed subtask in its pending slot.
- [PTO ISA TPush.hpp, shouldWaitFree](https://github.com/hw-native-sys/pto-isa/blob/5a4f74cbf627d4aac2e0ce10d5e0d8b118343265/include/pto/npu/a5/TPush.hpp#L58): initial slots skip free-credit waiting. The local no-split constructor does not establish consumer-ready initial credits.
- [PTO ISA no-split C2V transfer](https://github.com/hw-native-sys/pto-isa/blob/5a4f74cbf627d4aac2e0ce10d5e0d8b118343265/include/pto/npu/a5/TPush.hpp#L217): `AccToVecMode::SingleModeVec0` writes directly into the paired Vector UB.

In the original failing TP4 model, a level-4 trace and outputs from the **same timing pass** show the following timeline on AIC0 / AIV32 (microseconds relative to decoder start):

| Task | Core | Start | End |
| --- | --- | ---: | ---: |
| index-key decode | AIV32 | 0.000 | 57.186 |
| index weights Cube | AIC0 | 27.755 | 42.538 |
| index weights Vector | AIV32 | 58.022 | 59.867 |

The Cube producer finishes its push while the decoder still owns the paired Vector UB. All **96/96 corrupted model weight bit patterns** can be reconstructed by reinterpreting the damaged BF16 decoded-key bytes as FP32, then applying the weight epilogue's BF16 cast, 1/64 scaling, and BF16 cast. All 3072 visible scores differ before Top-K; the sort and selected indices correctly follow the corrupted device scores.

This is distinct from a missing GM tensor dependency or an inaccurate Top-K comparator. The two tasks are mathematically independent. It is also distinct from the bidirectional single-task stall in #2767 and the within-task GM-store ordering issue in #2745.

## Workaround and possible repair boundary

The model workaround orders index-key decode after weight projection completion. With unchanged mathematical kernels, three ordinary TP4 diagnostic runs and one profiled run matched all captured weights, keys, and scores. Source-level guards are being carried in [pypto-lib PR #1296](https://github.com/hw-native-sys/pypto-lib/pull/1296), covering shared C1A Full/Reindex helpers in prefill and decode. Decode has the same source pattern; the independently captured original failure is prefill.

A general repair must prevent unsafe partial MIX startup, or establish a consumer-ready handshake before a local C2V producer writes the Vector UB. Merely changing Top-K tolerances cannot repair memory corruption. Setting `require_sync_start` alone did **not** fix the original one-block weight task: the pinned orchestrator only applies that flag when `block_num > 1`.

Related: #2767, #2745 (different failure modes).
